File: sop-sdk/sdk-csharp/SDKCSharp/Client/OpenRequest.cs
using System;
using System.Web;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using SDKCSharp.Common;
using SDKCSharp.Utility;
using SDKCSharp.Response;

namespace SDKCSharp.Client
{
    public class OpenRequest
    {
        private OpenConfig openConfig;
        private OpenHttp openHttp;

        public OpenRequest(OpenConfig openConfig)
        {
            this.openConfig = openConfig;
            this.openHttp = new OpenHttp(openConfig);
        }

        /// <summary>
        /// 请求服务器
        /// </summary>
        /// <param name="url">url</param>
        /// <param name="requestForm">请求表单信息</param>
        /// <param name="header">请求头</param>
        /// <returns></returns>
        public string Request(string url, RequestForm requestForm, Dictionary<string, string> header)
        {
            Dictionary<string, string> form = requestForm.Form;
            List<UploadFile> files = requestForm.Files;
            if (files != null && files.Count > 0)
            {
                return this.openHttp.PostFile(url, form, header, files);
            }
            else
            {
                RequestMethod requestMethod = requestForm.RequestMethod;
                if (requestMethod == RequestMethod.GET)
                {
                    string query = this.BuildGetQueryString(form, requestForm.Charset);
                    if (!string.IsNullOrEmpty(query))
                    {
                        url = url + "?" + query;
                    }
                    return openHttp.Get(url, header);
                }
                return this.openHttp.PostJsonBody(url, JsonUtil.ToJSONString(form), header);
            }
        }

        public string BuildGetQueryString(Dictionary<string, string> form, Encoding charset)
        {
            StringBuilder queryString = new StringBuilder();
            Dictionary<string, string>.KeyCollection keys = form.Keys;
            int i = 0;
            foreach (string keyName in keys)
            {
                if (i++ > 0)
                {
                    queryString.Append("&");
                }
                queryString.Append(keyName).Append("=")
                        .Append(HttpUtility.UrlEncode(form[keyName], charset));
            }
            return queryString.ToString();
        }



        protected string CauseException(Exception e)
        {
            ErrorResponse result = SopSdkErrors.HTTP_ERROR;
            return JsonUtil.ToJSONString(result);
        }

    }

}

File: sop-sdk/sdk-java/src/main/java/com/gitee/sop/sdk/client/OpenRequest.java
package com.gitee.sop.sdk.client;

import com.alibaba.fastjson.JSON;
import com.gitee.sop.sdk.common.OpenConfig;
import com.gitee.sop.sdk.common.RequestForm;
import com.gitee.sop.sdk.common.RequestMethod;
import com.gitee.sop.sdk.common.SopSdkErrors;
import com.gitee.sop.sdk.common.UploadFile;
import com.gitee.sop.sdk.response.ErrorResponse;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.List;
import java.util.Map;

/**
 * 负责请求操作
 *
 * @author tanghc
 */
public class OpenRequest {

    private OpenHttp openHttp;

    public OpenRequest(OpenConfig openConfig) {
        this.openHttp = new OpenHttp(openConfig);
    }

    public String request(String url, RequestForm requestForm, Map<String, String> header) {
        try {
            Map<String, String> form = requestForm.getForm();
            List<UploadFile> files = requestForm.getFiles();
            if (files != null && files.size() > 0) {
                return openHttp.requestFile(url, form, header, files);
            } else {
                RequestMethod requestMethod = requestForm.getRequestMethod();
                if (requestMethod == RequestMethod.GET) {
                    String query = this.buildGetQueryString(form, requestForm.getCharset());
                    if (query != null && query.length() > 0) {
                        url = url + "?" + query;
                    }
                    return openHttp.get(url, header);
                } else {
                    return openHttp.request(url, form, header, requestMethod);
                }
            }
        } catch (IOException e) {
            return this.causeException(e);
        }
    }

    protected String buildGetQueryString(Map<String, String> params, String charset) throws UnsupportedEncodingException {
        if (params == null || params.size() == 0) {
            return "";
        }
        StringBuilder query = new StringBuilder();
        int i = 0;
        for (Map.Entry<String, String> entry : params.entrySet()) {
            String name = entry.getKey();
            String value = entry.getValue();
            if (i++ > 0) {
                query.append("&");
            }
            query.append(name).append("=").append(URLEncoder.encode(value, charset));
        }
        return query.toString();
    }

    protected String causeException(Exception e) {
        ErrorResponse result = SopSdkErrors.HTTP_ERROR.getErrorResponse();
        return JSON.toJSONString(result);
    }
}
